Firestore/core/src/firebase/firestore/api/query_core.mm

```diff
@@ -329,6 +324,101 @@ QuerySnapshot result(firestore_, query_, std::move(snapshot),
   }
 }
 
+void Query::ValidateDisjunctiveFilterElements(
+    const model::FieldValue& field_value,
+    const core::Filter::Operator op) const {
+  if (field_value.type() != FieldValue::Type::Array ||
+      field_value.array_value().size() == 0) {
+    ThrowInvalidArgument("Invalid Query. A non-empty array is required for '%s'"
+                         " filters.",
+                         Describe(op));
+  }
+  if (field_value.array_value().size() > 10) {
+    ThrowInvalidArgument("Invalid Query. '%s' filters support a maximum of 10"
+                         " elements in the value array.",
+                         Describe(op));
+  }
+
+  std::vector<FieldValue> array = field_value.array_value();
+  for (const auto& val : array) {
+    if (val.type() == FieldValue::Type::Null) {
+      ThrowInvalidArgument(
+          "Invalid Query. '%s' filters cannot contain 'null' in"
+          " the value array.",
+          Describe(op));
+    }
+    if ((val.type() == FieldValue::Type::Integer &&
+         std::isnan(val.integer_value())) ||
```

Comment:
This is nonsensical. Integers can't ever be NaN--that's a value that only exists in floating point numbers.

This happens to compile because of an implicit conversion.

Also, there's already a `FieldValue::is_nan()`.

There's also a `FieldValue::is_null()` you can use above.